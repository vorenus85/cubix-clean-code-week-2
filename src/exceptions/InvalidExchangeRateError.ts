export class InvalidExchangeRateError extends Error {
    public constructor(message: string) {
        super(message);
        this.name = this.constructor.name;
    }
}