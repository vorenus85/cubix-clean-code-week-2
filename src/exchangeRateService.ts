export interface IExchangeRateService {
    getExchangeRate(fromCurrency: string, toCurrency: string): number;
}