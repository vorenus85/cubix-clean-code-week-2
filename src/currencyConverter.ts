import { InvalidAmountInputError } from "./exceptions/InvalidAmountInputError";
import { InvalidExchangeRateError } from "./exceptions/InvalidExchangeRateError";
import { UnableToFetchExchangeRateError } from "./exceptions/UnableToFetchExchangeRateError";
import { IExchangeRateService } from "./exchangeRateService";

export class CurrencyConverter {

    private readonly FIXED_AMOUNT = 100;

    constructor(private exchangeRateService: IExchangeRateService) { }

    public Convert(amount: number, fromCurrency: string, toCurrency: string): number {
        this.validateAmount(amount);
        const exchangeRate = this.getExchangeRate(fromCurrency, toCurrency);
        this.validateExchangeRate(exchangeRate);
        return amount * exchangeRate;
    }

    public GenerateConversionReport(
        fromCurrency: string,
        toCurrency: string,
        startDate: Date,
        endDate: Date
    ): string {
        const conversions: number[] = [];

        const currentDate = new Date(startDate);

        while (currentDate <= endDate) {
            const exchangeRate = this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
            this.validateExchangeRate(exchangeRate);
            this.calculateConversion(exchangeRate, conversions, currentDate);
        }

        return `Conversion Report:\n${conversions.join('\n')}`;
    }

    private getExchangeRate(fromCurrency: string, toCurrency: string) {
        return this.exchangeRateService.getExchangeRate(fromCurrency, toCurrency);
    }

    private calculateConversion(exchangeRate: number, conversions: number[], currentDate: Date) {
        const convertedAmount = this.FIXED_AMOUNT * exchangeRate; // Assume a fixed amount for simplicity
        conversions.push(convertedAmount);
        currentDate.setDate(currentDate.getDate() + 1);
    }

    private validateExchangeRate(exchangeRate: number) {
        if (!exchangeRate) {
            throw new UnableToFetchExchangeRateError('Unable to fetch exchange rate.');
        }

        if (isNaN(exchangeRate)) {
            throw new InvalidExchangeRateError('Invalid exchange rate.');
        }
    }

    private validateAmount(amount: number) {
        if (isNaN(amount)) {
            throw new InvalidAmountInputError('Invalid amount input.');
        }
    }
}