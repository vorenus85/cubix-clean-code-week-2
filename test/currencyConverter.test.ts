import { InvalidAmountInputError } from "../src/exceptions/InvalidAmountInputError";
import { InvalidExchangeRateError } from "../src/exceptions/InvalidExchangeRateError";
import { UnableToFetchExchangeRateError } from "../src/exceptions/UnableToFetchExchangeRateError";

import { CurrencyConverter } from "../src/currencyConverter"
import { IExchangeRateService } from "../src/exchangeRateService";
import { mock, mockReset } from 'jest-mock-extended';

const mockExchangeRateService = mock<IExchangeRateService>();

describe('CurrencyConverter test::', ()=>{
    let currencyConverter: CurrencyConverter

    beforeEach(()=>{
        mockReset(mockExchangeRateService)
        currencyConverter = new CurrencyConverter(mockExchangeRateService)
    
    });

    describe('Happy scenarios', ()=>{
        it('should convert currency to another currency',()=>{
            // Arrange
            const amount = 1000;
            const fromCurrency = "HUF";
            const toCurrency = "USD";
            const mockedExchangeRate = 2;
            const expected = 2000;
    
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockReturnValue(mockedExchangeRate);
    
            // Act
            const actualResult = currencyConverter.Convert(amount, fromCurrency, toCurrency)
    
            // Assert
            expect(actualResult).toBe(expected);
        });

        it('should generate conversion report', ()=>{

            // Arrange
            const fromCurrency = "HUF";
            const toCurrency = "USD";
            const startDate = new Date('2023-01-01');
            const endDate = new Date('2023-01-05');
    
            const mockedExchangeRate = 2;
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockReturnValue(mockedExchangeRate);
    
            // Act
            const actualResult = currencyConverter.GenerateConversionReport(fromCurrency, toCurrency, startDate, endDate)
    
            // Assert
            expect(actualResult).toMatchSnapshot();
    
        });
    });

    describe('Error scenarios', ()=>{
        it('should throw error because of invalid exchangeRate',()=>{
            // Arrange
            const amount = 1000;
            const fromCurrency = "HUF";
            const toCurrency = "USD";
            const mockedExchangeRate = NaN;
            const errorMessage = 'Invalid exchange rate.';
            const expectedError = new InvalidExchangeRateError(errorMessage);
    
            // ensure the mock objectâ€™s internal state is verified?? :)
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockReturnValue(mockedExchangeRate).mockImplementation(() => { throw expectedError });
    
            // Act
    
            // Assert
            expect( ()=> currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedError);
        });
    
        it('should throw error because of invalid amount input',()=>{
            // Arrange
            const amount = NaN;
            const fromCurrency = "HUF";
            const toCurrency = "USD";
            const mockedExchangeRate = 2;
            const errorMessage = 'Invalid amount input.';
            const expectedError = new UnableToFetchExchangeRateError(errorMessage);
    
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockReturnValue(mockedExchangeRate);
    
            // Act
    
            // Assert
            expect( ()=> currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedError);
        });

        it('should throw error to because fromCurrency and toCurrency is empty ',()=>{
            // Arrange
            const amount = 1000;
            const fromCurrency = "";
            const toCurrency = "";
            const errorMessage = "Unable to fetch exchange rate.";
            const expectedError = new UnableToFetchExchangeRateError(errorMessage);
    
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockImplementation(()=> { throw expectedError} );
    
            // Assert
            expect( () => currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedError);
        });
    
        it('should throw error to because fromCurrency is empty ',()=>{
            // Arrange
            const amount = 1000;
            const fromCurrency = "";
            const toCurrency = "USD";
            const errorMessage = "Unable to fetch exchange rate.";
            const expectedError = new UnableToFetchExchangeRateError(errorMessage)
    
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockImplementation(()=> { throw expectedError} );
    
            // Assert
            expect( () => currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedError);
        });

        it('should throw error to because toCurrency is empty ',()=>{
            // Arrange
            const amount = 1000;
            const fromCurrency = "HUF";
            const toCurrency = "";
            const errorMessage = "Unable to fetch exchange rate.";
            const expectedError = new UnableToFetchExchangeRateError(errorMessage);
    
            mockExchangeRateService.getExchangeRate.calledWith(fromCurrency,toCurrency).mockImplementation(()=> { throw expectedError} );
    
            // Assert
            expect( () => currencyConverter.Convert(amount, fromCurrency, toCurrency)).toThrow(expectedError);
        });
    });
})