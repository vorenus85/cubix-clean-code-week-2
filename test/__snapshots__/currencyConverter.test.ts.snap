// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurrencyConverter test:: Happy scenarios should generate conversion report 1`] = `
"Conversion Report:
200
200
200
200
200"
`;

exports[`CurrencyConverter test:: should generate conversion report 1`] = `
"Conversion Report:
200
200
200
200
200"
`;
